pkg/compare/container.go

```diff
@@ -0,0 +1,140 @@
+	var out []byte
+	var err error
+	if engine.requiresSudo {
+		args = append([]string{engine.name}, args...) // Prepend engine name to args
+		out, err = execCommand("sudo", args...).CombinedOutput()
+	} else {
+		out, err = execCommand(engine.name, args...).CombinedOutput()
+	}
+
+	return out, err
+}
+
+// newEngine checks if Podman or Docker are in the system's PATH, and returns an engine with a name and a boolean
+// that indicates if sudo is needed for future commands. Prefers Podman. Returns an error if neither engine is found.
+var newEngine = func() (*engine, error) {
+	if _, err := lookPath("podman"); err == nil {
+		return &engine{name: "podman", requiresSudo: false}, nil
+	}
+
+	if _, err := lookPath("docker"); err == nil {
+		_, err = execCommand("docker", "images").Output() // If this errors out, we need to use sudo, return true.
+		return &engine{name: "docker", requiresSudo: err != nil}, nil
+	}
+	return &engine{name: "", requiresSudo: false}, fmt.Errorf("you do not have Podman or Docker on your PATH")
+}
+
+// pullContainer pulls an image, runs it using the provided engine, and stores the corresponding containerID in the engine struct
+func (engine *engine) pullAndRunContainer(image string) error {
+	// run because copy requires a running or stopped container
+	// -d to output container ID
+	out, err := engine.runEngineCommand("run", "-d", image)
+	if err != nil {
+		return fmt.Errorf("could not pull/run container: %s", out)
+	}
+	engine.containerID = strings.TrimSpace(string(out)) // Convert bytes to string and trim new line
+	return nil
+}
+
+// extractReferences copies the directory in the container that contains the reference configs into a temporary directory,
+// and stores the path to the new directory in the engine struct.
+func (engine *engine) extractReferences(pathToMetadata string, dname string) error {
+
+	out, err := engine.runEngineCommand("cp", engine.containerID+":"+pathToMetadata, dname)
+	if err != nil {
+		return fmt.Errorf("could not copy templates from container: %s", out)
+	}
+	engine.tempDir = filepath.Join(dname, filepath.Base(pathToMetadata))
+	return nil
+}
+
+// cleanup stops and removes the container used to extract the reference configs.
+func (engine *engine) cleanup() error {
+	out, err := engine.runEngineCommand("stop", engine.containerID)
+	if err != nil {
+		// Print errors as warnings rather than returning, since stopping and removing the container is not vital.
+		fmt.Printf("Warning: Could not stop container: %s", out)
+	}
+	out, err = engine.runEngineCommand("rm", engine.containerID)
+	if err != nil {
+		fmt.Printf("Warning: Could not remove container: %s", out)
+	}
+	return nil
+}
+
+// getReferencesFromContainer uses a path to an image and a metadata.yaml within that image, and extracts the reference configs
+// to a local temporary directory. Returns the path to this directory.
+func getReferencesFromContainer(path string, tempContainerRefDir string) (string, error) {
```

Comment:
Minor nit, but some code linters may complain about this - If you have 2 args of the same type in a row, you can combine them like this:
```suggestion
func getReferencesFromContainer(path, tempContainerRefDir string) (string, error) {
```